Fix CI

mikemhenry opened the pull request:
## Description
Provide a brief description of the PR's purpose here.

## Todos
- [ ] Implement feature / fix bug
- [ ] Add [tests](https://github.com/choderalab/openmmtools/tree/master/openmmtools/tests)
- [ ] Update [documentation](https://github.com/choderalab/openmmtools/tree/master/docs) as needed
- [ ] Update [changelog](https://github.com/choderalab/openmmtools/blob/master/docs/releasehistory.rst) to summarize changes in behavior, enhancements, and bugfixes implemented in this PR

## Status
- [ ] Ready to go

## Changelog message
```

```

ijpulidos commented:
Locally it's failing for me with scipy 1.14 but not with previous ones. Checking that in CI now.

ijpulidos commented:
And now it seems that `nose` is using the outdated methods in the `inspect` module, which has been deprecated for a while and removed in 3.11 as far as I can see. Seems like migrating to `pytest` is going to be required now!

ijpulidos commented:
Related issue https://github.com/choderalab/openmmtools/issues/661

mikemhenry commented:
Do we want to support windows? I am also not sure if this is a "real" error or not, but IMHO we should have an except for `ZeroDivisionError` and set `self._timing_data["ns_per_day"] = NaN` or something 
```
================================== FAILURES ===================================
______ TestHarmonicOscillatorsSAMSSampler.test_without_unsampled_states _______
[gw2] win32 -- Python 3.10.0 C:\Users\runneradmin\micromamba\envs\openmmtools-test\python.exe

self = <test_sampling.TestHarmonicOscillatorsSAMSSampler object at 0x000001A6B7C13880>

    def test_without_unsampled_states(self):
        """Test multistate sampler on a harmonic oscillator without unsampled endstates"""
>       self.run(include_unsampled_states=False)

openmmtools\tests\test_sampling.py:315: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
openmmtools\tests\test_sampling.py:199: in run
    simulation.run()
C:\Users\runneradmin\micromamba\envs\openmmtools-test\lib\site-packages\openmmtools\multistate\multistatesampler.py:798: in run
    self._update_timing(iteration_time, partial_total_time, run_initial_iteration, iteration_limit)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <instance of SAMSSampler>, iteration_time = 0.0, partial_total_time = 0.0
run_initial_iteration = 0, iteration_limit = 5000

    def _update_timing(self, iteration_time, partial_total_time, run_initial_iteration, iteration_limit):
        """
        Function that computes and transmits timing information to reporter.
    
        Parameters
        ----------
        iteration_time : float
            Time took in the iteration.
        partial_total_time : float
            Partial total time elapsed.
        run_initial_iteration : int
            Iteration where to start/resume the simulation.
        iteration_limit : int
            Hard limit on number of iterations to be run by the sampler.
        """
        self._timing_data["iteration_seconds"] = iteration_time
        self._timing_data["average_seconds_per_iteration"] = \
            partial_total_time / (self._iteration - run_initial_iteration)
        estimated_timedelta_remaining = datetime.timedelta(
            seconds=self._timing_data["average_seconds_per_iteration"] * (iteration_limit - self._iteration)
        )
        estimated_finish_date = datetime.datetime.now() + estimated_timedelta_remaining
        self._timing_data["estimated_time_remaining"] = str(estimated_timedelta_remaining)  # Putting it in dict as str
        self._timing_data["estimated_localtime_finish_date"] = estimated_finish_date.strftime("%Y-%b-%d-%H:%M:%S")
        total_time_in_seconds = datetime.timedelta(
            seconds=self._timing_data["average_seconds_per_iteration"] * iteration_limit
        )
        self._timing_data["estimated_total_time"] = str(total_time_in_seconds)
    
        # Estimate performance
        moves_iterator = self._flatten_moves_iterator()
        # Only consider "dynamic" moves (timestep and n_steps attributes)
        moves_times = [move.timestep.value_in_unit(unit.nanosecond) * move.n_steps for move in moves_iterator if
                       hasattr(move, "timestep") and hasattr(move, "n_steps")]
        iteration_simulated_nanoseconds = sum(moves_times)
        seconds_in_a_day = (1 * unit.day).value_in_unit(unit.seconds)
>       self._timing_data["ns_per_day"] = iteration_simulated_nanoseconds / (
                    self._timing_data["average_seconds_per_iteration"] / seconds_in_a_day)
E       ZeroDivisionError: float division by zero

C:\Users\runneradmin\micromamba\envs\openmmtools-test\lib\site-packages\openmmtools\multistate\multistatesampler.py:1779: ZeroDivisionError
```

ijpulidos commented:
I'd vote for making windows support not required. As in, we can make the windows CI job optional.

ijpulidos commented:
More importantly, I think we are not running a big chunk of the tests when changing to pytest. Locally I'm getting the following:

* With nose on main branch: `Ran 2044 tests in 2222.346s`
* With pytest (this branch): `342 passed, 14 skipped, 1 xfailed, 1495 warnings in 973.57s (0:16:13)`

I'm not sure `pytest` and `nose` count tests in the same way but I think this is something we should double check before merging these changes. I can dig into which tests exactly are missing, but maybe related to that `xfailed` one?

ijpulidos commented:
> Do we want to support windows? I am also not sure if this is a "real" error or not, but IMHO we should have an except for `ZeroDivisionError` and set `self._timing_data["ns_per_day"] = NaN` or something

Hmm, good question. I don't know if by making it catch the `ZeroDivisionError` we would be masking other errors, but definitely something to think about.

mikemhenry commented:
```
---------- coverage: platform linux, python 3.11.0-final-0 -----------
Name                                           Stmts   Miss  Cover
------------------------------------------------------------------
openmmtools/__init__.py                            8      8     0%
openmmtools/alchemy/__init__.py                    1      1     0%
openmmtools/alchemy/alchemy.py                   864    864     0%
openmmtools/cache.py                             288    288     0%
openmmtools/constants.py                          11     11     0%
openmmtools/forcefactories.py                     69     69     0%
openmmtools/forces.py                            361    361     0%
openmmtools/integrators.py                       764    764     0%
openmmtools/mcmc.py                              463    463     0%
openmmtools/multistate/__init__.py                 7      7     0%
openmmtools/multistate/multistateanalyzer.py    1016   1016     0%
openmmtools/multistate/multistatereporter.py     712    712     0%
openmmtools/multistate/multistatesampler.py      735    735     0%
openmmtools/multistate/paralleltempering.py       57     57     0%
openmmtools/multistate/pymbar.py                  25     25     0%
openmmtools/multistate/replicaexchange.py        118    118     0%
openmmtools/multistate/sams.py                   287    287     0%
openmmtools/multistate/utils.py                   68     68     0%
openmmtools/respa.py                              33     33     0%
openmmtools/scripts/__init__.py                    0      0   100%
openmmtools/scripts/test_openmm_platforms.py     257    257     0%
openmmtools/sobol.py                             159    159     0%
openmmtools/states.py                           1188   1188     0%
openmmtools/storage/__init__.py                    2      2     0%
openmmtools/storage/iodrivers.py                 713    713     0%
openmmtools/storage/storageinterface.py          138    138     0%
openmmtools/testsystems.py                      1422   1422     0%
openmmtools/utils/__init__.py                      2      2     0%
openmmtools/utils/equilibration.py                74     74     0%
openmmtools/utils/utils.py                       361    361     0%
------------------------------------------------------------------
TOTAL                                          10203  10203     0%
Coverage XML written to file coverage.xml
```

I've seen this before, I think it has to do with not doing a dev install

mikemhenry commented:
it worked! https://github.com/choderalab/openmmtools/actions/runs/10567725972/job/29277367842

mikemhenry commented:
slow tests running here: https://github.com/choderalab/openmmtools/actions/runs/10568140366/job/29278645522

ijpulidos commented:
@mikemhenry nice! We would need to separate the ruff/black/formatter changes in a separate commit, that way it's going to be way easier for me to review, and hopefully also ignore those so that we don't change the blame, which I think it's important to maintain.

mikemhenry commented:
@ijpulidos when we merge this in, it will be squashed into a single commit we can add it to `.git-blame-ignore-revs` https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view 

> We would need to separate the ruff/black/formatter changes in a separate commit

I will see if I can rebase things that way

mikemhenry commented:
Or maybe it would be better to squash everything into two commits, automated refactor and non-automated, and only add the auotmated refactor commit to `.git-blame-ignore-revs` and merge in the pr instead of sqash and merge

ijpulidos commented:
> Or maybe it would be better to squash everything into two commits, automated refactor and non-automated, and only add the auotmated refactor commit to `.git-blame-ignore-revs` and merge in the pr instead of sqash and merge

Yes, that would definitely make the review easier. Considering it's such a big set of changes otherwise, I would just review the non-automated commit.

mikemhenry commented:
Okay I've rebased a few commits on a different branch (If I made a mistake I didn't want to wreck all of my work). Once you are good with it, I will add a `.git-blame-ignore-revs` and merge in that branch.

Commits to review:
https://github.com/choderalab/openmmtools/commit/c710d687e8c4d536967e2a10ed7dadc836ff52ce
https://github.com/choderalab/openmmtools/commit/c8882e1098a2987b3615762d2f109e5f692563d2
https://github.com/choderalab/openmmtools/commit/be3287bd3cf190aea684d5266f72052320075942
https://github.com/choderalab/openmmtools/commit/c5c9acc593b3e659c4f961d67bdeaed1ba863cd6

Commits to add to `.git-blame-ignore-revs` (automated formatting stuff)
https://github.com/choderalab/openmmtools/commit/562e0d305b96082cac337ce2320f5f7835385618
https://github.com/choderalab/openmmtools/commit/7ebe1ab6e86582e8ed5e2cc2103d8eaa599ec3f2
https://github.com/choderalab/openmmtools/commit/e7144c193ad307aa9ef2f852feecf0dcdd2ef352 # this one is where I updated versioneer

ijpulidos commented on openmmtools/tests/test_sampling.py:
```diff
@@ -1645,10 +2175,14 @@ class TestReplicaExchange(TestMultiStateSampler):
         self.actual_stored_properties_check(additional_properties=additional_values)
 
-    @attr('slow')  # Skip on Travis-CI
+    @pytest.mark.slow  # Skip on Travis-CI
```
Non-blocking comment. Just for future references, there are a few of tests that are marked as slow. We should probably try to reassess if these are still slow with modern computers/CPUs.

ijpulidos replied:
Created an issue for this as well. https://github.com/choderalab/openmmtools/issues/748

ijpulidos commented on openmmtools/tests/test_storage_interface.py:
```diff
@@ -42,27 +44,27 @@ def spawn_driver(path):
         driver = spawn_driver(test_store)
         si = StorageInterface(driver)
-        si.var1.read()
+        with pytest.raises(KeyError):
```
Great we capturing the actual `KeyError` exception here, instead of a generic `Exception`.

ijpulidos commented on openmmtools/tests/test_utils.py:
```diff
@@ -480,61 +544,113 @@ def test_gentle_equilibration_setup(self):
 
     # TODO: Marking as not a test until we solve our GPU CI
-    @nottest
+    @pytest.mark.slow
```
Yes! We have a GPU CI now. Good catch!

ijpulidos commented on .github/workflows/CI.yml:
```diff
@@ -17,23 +17,22 @@ concurrency:
-        openmm: ["7.7", "8.0"]
-        os: [macOS-latest, ubuntu-latest]
+        python-version: ["3.10", "3.11", "3.12"]
+        openmm: ["8.0", "8.1.1", "8.1.2"]
```
Nonblocking comment. It's a great idea that we are testing with different OpenMM versions. I think for the longer term we should make sure we understand what kind of versioning system (semantic or not, etc.) is OpenMM using, such that we don't have to be testing with every minor OpenMM release out there. I hope this makes sense.

mikemhenry replied:
Yes, I think post this merge we can probably drop 8.0, and make a decision if we want to support n, n-1 openmm releases or something else.

ijpulidos replied:
Created an issue for this in https://github.com/choderalab/openmmtools/issues/747

ijpulidos commented on devtools/conda-envs/test_env.yaml:
```diff
@@ -4,7 +4,7 @@ channels:
     # Base depends
   - cython
-  - hdf5 <=1.14.0  # Macos has problem with newer releases (1.14.x) to date
+    #- hdf5 <=1.14.0  # Macos has problem with newer releases (1.14.x) to date
```
I do not fully remember why we had to pin the version at the time, but this was probably something that was not fixed upstream (with the python bindings/wrappers). Good to see it's working without pinning!!

ijpulidos commented on devtools/conda-envs/test_env.yaml:
```diff
@@ -18,15 +18,14 @@ dependencies:
   - python
   - pyyaml
-  - scipy
+  - scipy <1.14
```
Non-blocking comment. Why is it that we are not supporting scipy 1.14+? I wonder if we then have to restrict the version on the conda-forge package and also patch/change all the currently released packages so that older versions don't get installed when environment with scipy 1.14 are resolved. I hope that makes sense.

mikemhenry replied:
I will relax the pin and see if we have any issues, I am pretty sure it was actually something to do with nose testing or a numba thing

ijpulidos replied:
Yes, that should be a quick test, great!

ijpulidos replied:
Oh, I think it was probably related to this, which should've been already fixed on the openmm side. But if we are supporting older OpenMM releases then we might still need the pin. https://github.com/choderalab/openmmtools/issues/743

mikemhenry replied:
we will need the pin until this gets merged in https://github.com/conda-forge/openmm-feedstock/pull/139 + we patch the older releases

ijpulidos commented on .github/workflows/gpu-runner.yaml:
```diff
@@ -1,6 +1,9 @@
   workflow_dispatch:
+  schedule:
+    # At 07:00 UTC daily.
+    - cron: "0 7 * * *"
```
Do we want to run daily? I think we should be okay if we run weekly, but I don't have strong feelings about this. Just thinking about mitigating costs (even if low).

mikemhenry replied:
I remember thinking the same thing, but John pointed out it doesn't cost much and it can be nice to track down when an upstream thing changes and only have to parse through a env diff between days instead of weeks.

ijpulidos reviewed (approved):
This is great! Finally we can move to a modern unit testing infrastructure and GPU CI. There are just minor non-blocking comments.

I think we have talked about how `nose` counts tests differently compared to `pytest` and that the coverage information should account for this. We want to make sure we are doing as good or better in terms of coverage.

Great work! LGTM!

mikemhenry commented:
Done in #746
